app/models/language.rb

```diff
@@ -58,55 +61,116 @@ def self.menu_options
+  def arel_select_top_contributors_in_order(user_ids)
+    u = User.arel_table
+    u.project([u[:id], u[:login]]).where(u[:id].in(user_ids))
+  end
+
+  public
+
   # Count the number of lines the user has translated.  Include edits, as well.
   # It counts paragraphs, actually, and weights them according to length.
   def self.calculate_users_contribution(user)
     lines = 0
-    for text in Language.connection.select_values(%(
-      SELECT GROUP_CONCAT(CONCAT(text, "\n")) AS x
-      FROM translation_strings_versions
-      WHERE user_id = #{user.id}
-      GROUP BY translation_string_id
-    ))
+    select_manager = arel_select_translation_strings_versions(user)
+    # puts(select_manager.to_sql)
+    values = Language.connection.select_values(select_manager.to_sql)
+    # # puts(values.inspect)
+    for text in values
       lines += score_lines(text)
     end
     lines
   end
 
+  # SELECT GROUP_CONCAT(CONCAT(text, "\n")) AS x
+  # FROM translation_strings_versions
+  # WHERE user_id = #{user.id}
+  # GROUP BY translation_string_id
+  private_class_method def self.arel_select_translation_strings_versions(user)
+    v = Arel::Table.new(:translation_strings_versions)
+    v.where(v[:user_id].eq(user.id)).
+      group(v[:translation_string_id]).
+      project(arel_function_group_concat_strings(v))
+  end
+
   def calculate_users_contribution(user)
     lines = 0
-    for text in Language.connection.select_values(%(
-      SELECT GROUP_CONCAT(CONCAT(v.text, "\n")) AS x
-      FROM translation_strings t, translation_strings_versions v
-      WHERE t.language_id = #{id}
-        AND v.translation_string_id = t.id
-        AND v.user_id = #{user.id}
-      GROUP BY t.id
-    ))
+    select_manager = arel_select_translation_strings(user)
+    # puts(select_manager.to_sql)
+    values = Language.connection.select_values(select_manager.to_sql)
+    # puts(values.inspect)
+    for text in values
       lines += Language.score_lines(text)
     end
     lines
   end
 
+  private
+
+  # SELECT GROUP_CONCAT(CONCAT(v.text, "\n")) AS x
+  # FROM translation_strings t, translation_strings_versions v
+  # WHERE t.language_id = #{id}
+  #   AND v.translation_string_id = t.id
+  #   AND v.user_id = #{user.id}
+  # GROUP BY t.id
+  def arel_select_translation_strings(user)
+    t = Arel::Table.new(:translation_strings)
+    v = Arel::Table.new(:translation_strings_versions)
+    TranslationString.joins(t.join(v).on(t[:language_id].eq(id).
+                            and(v[:translation_string_id].eq(t[:id])).
+                            and(v[:user_id].eq(user.id))).join_sources).
+      group(t[:id]).select(self.class.arel_function_group_concat_strings(v))
+  end
+
+  public
+
+  # JASON QUESTION - Is there a way to rewrite these without the group_concat?
+  # I was thinking maybe just select the :text, and append the newlines in Ruby
+  # via map. But this gives a different score, because of the grouping.
+  # The command in PostgresQL is "STRING_AGG"
```

Comment:
I hear you, Nimmo.  And I am twitching to massively simplify this in some way, too.  I feel like we could still do the grouping in ruby...

Okay, I'm understanding the scoring a bit better.  The tricky part is that we don't want to count every line of long translation strings for each tiny change.  Say, for example, you have 100 line help message as a single translation string.  Now you add a single word.  We want to count that as only a single line change, not 100!  That's what the hashifying business is in Language.score_lines.

Okay, so what if we returned pairs of translation_string_id and text instead of just text, without the grouping.  Then score_lines could include the translation_string_id in the hash key, and I think it might turn out equivalent?

Does this look right?  I wonder how it would perform, mashing everything into one enormous hash like that?

```suggestion
  # This one is not used, rip it out!
  # def self.calculate_users_contribution(user)
  # end

  def calculate_users_contribution(user)
    hash = {}
    TranslationString.joins(:versions).where(id: id, versions: { user: user }).pluck(:id, "versions.text").each do |id, text|
      text.split("\n").each { |str| hash[id, str] = true if str.present? }
    end
    score_lines(hash)
  end

  def score_lines(hash)
    score = 0
    hash.each_key do |id, str|
      score += (key.length.to_f / CHARACTERS_PER_LINE).truncate + 1
    end
    score
  end
```

Reply:
Note : `Language#self.calculate_users_contribution(user)` is used in `language_test.rb`, we need to delete there also, if the class method is deleted.

Reply:
I believe your line 168 should be 
```ruby
TranslationString.joins(:versions).where(language_id: id, versions: { user_id: user.id }).pluck(etc
```

Reply:
Ok. I realize the above is just sketched out, but something's off and I can't figure out what it is. 

At least, it's calculating higher scores than before:

Language 4
user 55281 "MO Webmaster"
score before: 18
score after: 65

Now, the webmaster seems to have mostly translated flash messages, which are long. But they all have the same ID. There are 14 strings for ID 14040, like these first two:
```
[14040, "Be a part of Mushroom Observer’s big plans! <br>\r\n<br>\r\nPlease consider \"**+donating+**\":/support/donate before the end of 2020 to help MO determine how many of our \"**+ideas+**\":/articles/16 we can implement over the next year.<br>\r\n<br>\r\nThis year only U.S. taxpayers who won’t be itemizing deductions for 2020 can get a \"**+special deduction+**\":/articles/17 of up to $300 for charitable contributions made before the end of 2020."], 
[14040, "Do you want to design a logo for Mushroom Observer? <br>\r\nWe are running a contest!  Please consider offering an idea. \r\n<a href=\"https://docs.google.com/document/d/1y9OwJhX8eD6FIkrbIXah3gXRMyFYrox0PE-q52R-YHc/edit?usp=sharing\">Click here for contest rules</a>"]
```

Reply:
Definitely delete the test, too.  No reason to have (or test!) code which isn't used.
I need to look at this more carefully, too.  You're right, it was a sketch while very tired!

Reply:
Okay, I think I see the problem.  It's counting multiple versions of the same translation too many times.  Try:

```suggestion
  def calculate_users_contribution(user)
    hash = {}
    TranslationString.joins(:versions).where(id: id, versions: { user: user }).pluck(:id, "versions.text").each do |id, text|
      text.split("\n").each { |str| (hash[id] ||= {}) && (hash[id][str] = true) if str.present? }
    end
    score_lines(hash)
  end
  
  def score_lines(hash)
    score = 0
    hash.each_value do |hash2|
      hash2.each_key do |key|
        score += (key.length.to_f / CHARACTERS_PER_LINE).truncate + 1
      end
    end
    score
  end
```

Now it is truly doing the original "group" entirely in ruby.

Reply:
Thanks. I tried this, and it actually still produces the same higher score, "65", for MO Webmaster. I also tried grouping the AR result, but that produced a **lower** than original score, "8".

Here's what i'm using, note the small syntax tweaks in case you want to run locally:
```suggestion
  def calculate_users_contribution(user)
    # Note: Rails didn't like "versions.text"
    v = Arel::Table.new(:translation_strings_versions)
    hash = {}
    strings = TranslationString.joins(:versions).
              where(language_id: id,
                    translation_strings_versions: { user_id: user.id }). # full table name neeeded
              group(:id).pluck(:id, v[:text]) # without the `group` it produces much higher score
    strings.each do |id, text|
      text.split("\n").each do |str|
        (hash[id] ||= {}) && (hash[id][str] = true) if str.present?
      end
    end
    score_lines(hash)
  end

  def score_lines(hash)
    score = 0
    hash.each_value do |hash2|
      hash2.each_key do |key|
        score += (key.length.to_f / CHARACTERS_PER_LINE).truncate + 1
      end
    end
    score
  end
```